File: packages/compiler/src/expr.ts
function scanLogic(str: string, onTopLevel?: (char: string, index: number) => boolean): { depth: number, index: number } {
  let depthParen = 0
  let depthBracket = 0
  let depthBrace = 0
  const quoteStack: Array<'"' | "'" | '`' | '${'> = []
  let escaped = false

  for (let i = 0; i < str.length; i++) {
    const char = str[i]
    const quote = quoteStack[quoteStack.length - 1]

    if (quote === '"' || quote === "'" || quote === '`') {
      if (escaped) {
        escaped = false
        continue
      }
      if (char === '\\') {
        escaped = true
        continue
      }
      if (char === quote) {
        quoteStack.pop()
      } else if (quote === '`' && char === '$' && str[i + 1] === '{') {
        quoteStack.push('${')
        i++
      }
      continue
    }

    if (char === '"' || char === "'" || char === '`') {
      quoteStack.push(char)
      continue
    }

    if (char === '(') depthParen++
    else if (char === ')') depthParen = Math.max(0, depthParen - 1)
    else if (char === '[') depthBracket++
    else if (char === ']') depthBracket = Math.max(0, depthBracket - 1)
    else if (char === '{') {
      depthBrace++
    } else if (char === '}') {
      if (depthBrace === 0 && quote === '${') {
        quoteStack.pop()
      } else {
        depthBrace = Math.max(0, depthBrace - 1)
      }
    }

    if (
      onTopLevel &&
      depthParen === 0 &&
      depthBracket === 0 &&
      depthBrace === 0 &&
      quoteStack.length === 0 &&
      onTopLevel(char, i)
    ) {
      return { depth: 0, index: i }
    }
  }

  return { depth: depthBrace + depthParen + depthBracket + quoteStack.length, index: -1 }
}

export function scanTopLevel(str: string, onTopLevel: (char: string, index: number) => boolean): number {
  return scanLogic(str, onTopLevel).index
}

export function findTopLevelWhitespace(str: string): number {
  return scanTopLevel(str, (char) => /\s/.test(char))
}

export function findTopLevelColon(str: string): number {
  return scanTopLevel(str, (char) => char === ':')
}

export function findTopLevelEquals(str: string): number {
  return scanTopLevel(
    str,
    (char, index) =>
      char === '=' &&
      str[index + 1] !== '=' &&
      str[index + 1] !== '>' && // Skip arrow functions (=>)
      str[index - 1] !== '=' && // Skip compound (+=, -=, etc)
      str[index - 1] !== '!' &&
      str[index - 1] !== '<' &&
      str[index - 1] !== '>',
  )
}

export function unwrapBalancedBraces(str: string): string | null {
  const trimmed = str.trim()
  if (!trimmed.startsWith('{') || !trimmed.endsWith('}')) return null

  // We want to see if the first '{' is closed exactly at the end.
  // We can't use scanLogic directly because it treats the first '{' as depth 1.
  
  let depthBrace = 0
  const quoteStack: Array<'"' | "'" | '`' | '${'> = []
  let escaped = false

  for (let i = 0; i < trimmed.length; i++) {
    const char = trimmed[i]
    const quote = quoteStack[quoteStack.length - 1]

    if (quote === '"' || quote === "'" || quote === '`') {
      if (escaped) {
        escaped = false
        continue
      }
      if (char === '\\') {
        escaped = true
        continue
      }
      if (char === quote) {
        quoteStack.pop()
      } else if (quote === '`' && char === '$' && trimmed[i + 1] === '{') {
        quoteStack.push('${')
        i++
      }
      continue
    }

    if (char === '"' || char === "'" || char === '`') {
      quoteStack.push(char)
      continue
    }

    if (char === '{') {
      depthBrace++
    } else if (char === '}') {
      if (depthBrace === 0 && quote === '${') {
        quoteStack.pop()
      } else {
        depthBrace--
      }
    }

    if (depthBrace === 0 && i < trimmed.length - 1) {
      return null
    }
  }

  return depthBrace === 0 ? trimmed.slice(1, -1).trim() : null
}

File: packages/loom_core/src/expr.rs
pub fn scan_top_level<F>(str: &str, mut on_top_level: F) -> isize
where
    F: FnMut(char, usize) -> bool,
{
    let mut depth_paren = 0;
    let mut depth_bracket = 0;
    let mut depth_brace = 0;
    let mut quote: Option<char> = None;
    let mut escaped = false;

    for (i, c) in str.char_indices() {
        if let Some(q) = quote {
            if escaped {
                escaped = false;
                continue;
            }
            if c == '\\' {
                escaped = true;
                continue;
            }
            if c == q {
                quote = None;
            }
            continue;
        }

        if c == '"' || c == '\'' || c == '`' {
            quote = Some(c);
            continue;
        }

        match c {
            '(' => depth_paren += 1,
            ')' => depth_paren = std::cmp::max(0, depth_paren - 1),
            '[' => depth_bracket += 1,
            ']' => depth_bracket = std::cmp::max(0, depth_bracket - 1),
            '{' => depth_brace += 1,
            '}' => depth_brace = std::cmp::max(0, depth_brace - 1),
            _ => {}
        }

        if depth_paren == 0 && depth_bracket == 0 && depth_brace == 0 && on_top_level(c, i) {
            return i as isize;
        }
    }

    -1
}

pub fn find_top_level_whitespace(str: &str) -> isize {
    scan_top_level(str, |c, _| c.is_whitespace())
}

pub fn find_top_level_colon(str: &str) -> isize {
    scan_top_level(str, |c, _| c == ':')
}

pub fn find_top_level_equals(str: &str) -> isize {
    let chars: Vec<char> = str.chars().collect();
    let mut depth_paren = 0;
    let mut depth_bracket = 0;
    let mut depth_brace = 0;
    let mut quote: Option<char> = None;
    let mut escaped = false;

    for i in 0..chars.len() {
        let c = chars[i];

        if let Some(q) = quote {
            if escaped {
                escaped = false;
                continue;
            }
            if c == '\\' {
                escaped = true;
                continue;
            }
            if c == q {
                quote = None;
            }
            continue;
        }

        if c == '"' || c == '\'' || c == '`' {
            quote = Some(c);
            continue;
        }

        match c {
            '(' => depth_paren += 1,
            ')' => depth_paren = std::cmp::max(0, depth_paren - 1),
            '[' => depth_bracket += 1,
            ']' => depth_bracket = std::cmp::max(0, depth_bracket - 1),
            '{' => depth_brace += 1,
            '}' => depth_brace = std::cmp::max(0, depth_brace - 1),
            _ => {}
        }

        if depth_paren == 0 && depth_bracket == 0 && depth_brace == 0 && c == '=' {
            let next = chars.get(i + 1);
            let prev = if i > 0 { chars.get(i - 1) } else { None };

            let is_assignment = next != Some(&'=')
                && next != Some(&'>')
                && prev != Some(&'!')
                && prev != Some(&'<')
                && prev != Some(&'>')
                && prev != Some(&'=');

            if is_assignment {
                // Return byte offset
                return str.char_indices().nth(i).unwrap().0 as isize;
            }
        }
    }

    -1
}

pub fn unwrap_balanced_braces(str: &str) -> Option<String> {
    let trimmed = str.trim();
    if !trimmed.starts_with('{') || !trimmed.ends_with('}') {
        return None;
    }

    let mut depth = 0;
    let mut quote: Option<char> = None;
    let mut escaped = false;

    for (i, c) in trimmed.char_indices() {
        if let Some(q) = quote {
            if escaped {
                escaped = false;
                continue;
            }
            if c == '\\' {
                escaped = true;
                continue;
            }
            if c == q {
                quote = None;
            }
            continue;
        }

        if c == '"' || c == '\'' || c == '`' {
            quote = Some(c);
            continue;
        }

        if c == '{' {
            depth += 1;
        } else if c == '}' {
            depth -= 1;
        }

        if depth == 0 && i < trimmed.len() - 1 {
            return None;
        }
    }

    if depth == 0 {
        Some(trimmed[1..trimmed.len() - 1].trim().to_string())
    } else {
        None
    }
}
